### ISSUE TITLE
Template literal string types in Record not showing error

### ISSUE THREAD

**Issue by orlahowell** [microsoft/TypeScript#57272]
### 🔎 Search Terms

Template literal type as key in Record not enforcing type/throwing error when in function

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about string literals

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.3#code/GYVwdgxgLglg9mABFApgZygMXBAFAQwC5EAlFCOAJwBMAeAA2DjgH0ASAbw0pjAHMAvvQA0iAEbMANinxgAfAEpEHAFCJElFFBCUk+ANwqBKlRTAZEcMQCtEAXmVrETVmPyViUSiBSiA9H4AtvhQEAAW6BrkVNTIAJ4ADihObh7I3r6IAYjUcJFgcFCIwaFhURQ08UlGJiqoGNiQuFbWCvpZfohoYXAgkrHdcADuiCiUlFSIALRVKJY2iDBoiAVF+GhoMHxg+GLSyHCI9GQVdIzM7FxevIIi4lIy8vRAA

### 💻 Code

```ts
function testFunc(a: Record<`foo_${string}`, boolean>) {
  return a;
}

const obj = {
  foo_bar: true, //matches record type
  bar: true, // does not match record type
}


testFunc(obj); // should show error - type obj is not assignable to `Record<`foo_${string}`, boolean>`
```

### 🙁 Actual behavior

`testFunc` should show an error because `obj` is does not match the Record type.

### 🙂 Expected behavior

I would expect an error along the lines of `obj` is not assignable to `Record<`foo_${string}`, boolean>`, because `bar` is not assignable to `foo_${string}`.

**Comment by MartinJohns** [comment]
This is working as intended.

Objects are not sealed, they can have additional properties. All your record says is that all properties named starting with `foo_` must be of type `boolean`, and that's the case with the object you pass.
